templated BM  - MOD-3928 (#266)
RedisAI/VectorSimilarity: tests/benchmark/README.md

# Vector Similarity benchmarks

## Table of contents
* [Overview](#overview)
* [Run benchmarks](#run-benchmarks)
* [Available sets](#available-sets)  
    - [BM_VecSimBasics](#bm_vecsimbasics)  
    - [BM_BatchIterator](#bm_batchiterator)  
    - [BM_VecSimUpdatedIndex](#bm_vecsimupdatedindex)  
* [The benchmarks directory](#the-benchmarks-directory)
    - [How to define and register a new test](#how-to-define-and-register-a-new-test)
* [Fixtures](#fixtures)
    - [Benchmarks fixture classes hierarchy](#benchmarks-fixture-classes-hierarchy)
* [Cmake remarks](#cmake-remarks)
* [ann-benchmarks](#ann-benchmark)

# Overview
We use Google benchmark tool to create our benchmarks.

> Google benchmark is a popular tool for benchmark code snippets, similar to unit tests. It allows a quick way of estimating the runtime of each test based on several (identical) runs, and print the results as output. For some tests, the output includes additional "Recall" metric, which indicates the accuracy in case of approximate search.  

Author: why google benchmarks is unlike unit tests? :thinking:

**The recall** is calculated as the number of results from the approximate algorithm (HNSW in this case) divided by the ground truth results number (calculated by the brute force algorithm).

# Run benchmarks
## Required files
The serialized hnsw indices and the datasets for the ann-benchmarks can be found in 
`tests/benchmark/data/hnsw_indices.txt`.  
To download all the required files run:
```asm
wget -q -i tests/benchmark/data/hnsw_indices.txt -P tests/benchmark/data
```
To run all test sets, call the following commands from the project root dir:
```asm
make benchmark
```
### Running a Subset of Benchmarks
To run only a subset of benchmarks that match a specified `<regex>`, set `BENCHMARK_FILTER=<regex>` environment variable. For example:  
```asm
make benchmark BENCHMARK_FILTER=fp32*
```

# Available sets
There are currently 3 sets of benchmarks available: `BM_VecSimBasics` , `BM_BatchIterator` and `BM_VecSimUpdatedIndex`. Each is templated according to the index data type.
## BM_VecSimBasics
For each combination of data type (fp32/fp64) and label type (single/multi) the following test cases are included:
1. Index `total memory` 

> 2. `AddLabel` - adds `DEFAULT_BLOCK_SIZE (= 1024)` new labels to the index from the test_query_file. Note that for a single value index each label contains one vector, meaning that the number of new labels equals the number of new vectors.  

Author: i didn't mention any specific attributes of the datasets to leave the option to change them in the future.
I'll add a section describing all the current datasets' parameters so it will be written in one place and easy to change if needed.

**results:** average time per label, average memory addition per vector and vectors per label.  
*At the end of the benchmark, we delete the added labels*
3. `DeleteLabel` - Deletes `DEFAULT_BLOCK_SIZE (= 1024)` labels from the index. Note that for a single value index each label contains one vector, meaning that the number of deleted labels equals the number of deleted vectors.  
**results:** average time per label, average memory addition per vector (a negative value means that the memory has decreased).  

> *In each iteration , before deletion, we pause the time measurement to store the deleted label data so we can re-add it to the index at the end of the benchmark*

Author: its purpose is to say that it shouldn't affect the following benchmarks and that the re-adding time is not taken into account in the time results
still redundant?

> *In each iteration , before deletion, we pause the time measurement to store the deleted label data so we can re-add it to the index at the end of the benchmark*

Reviewer: Just the part about pausing the time. you can just leave: "*At the end of the benchmark, we re-add the deleted vectors under the same labels*

4. Run `Top_K` query over the flat index (using brute-force search), for each `k=10`, `k=100` and `k=500`  

> **results:** average time per iteration

Author: its not fixed. Usually 3-5 iterations

5. Run **100** iterations of `Top_K` query over the HNSW index, for each pair of `{ef_runtime, k}` from the following: 
    `{ef_runtime=10, k=10}`   
    `{ef_runtime=200, k=10}`   
    `{ef_runtime=100, k=100}`   
    `{ef_runtime=200, k=100}`   
    `{ef_runtime=500, k=500}`   
**results:** average time per iteration, Recall  
6. Run `Range` query over the flat index (using brute-force search), for each `radiusX100=20`, `radiusX100=35` and `radiusX100=50`  
**results:** average time per iteration, results number
7. Run **100** iterations of `Range` query over the HNSW index, for each pair of `{radiusX100, epsilonX1000}` from the following: 
    `{radiusX100=20, epsilonX1000=1}`   
    `{radiusX100=20, epsilonX1000=10}`   
    `{radiusX100=20, epsilonX1000=100}`   
    `{radiusX100=35, epsilonX1000=1}`   
    `{radiusX100=35, epsilonX1000=10}`   
    `{radiusX100=35, epsilonX1000=100}`   
    `{radiusX100=50, epsilonX1000=1}`   
    `{radiusX100=50, epsilonX1000=10}`   

>     `{radiusX100=50, epsilonX1000=100}`   

Author: its Beni Goren to me

> **results:** average time per iteration, results number, Recall  

Author: is "average results number per iteration" better?

`{batch size=10, number of batches=1}`  
`{batch size=10, number of batches=3}`  
`{batch size=10, number of batches=5}`  
`{batch size=100, number of batches=1}`  
`{batch size=100, number of batches=3}`  
`{batch size=100, number of batches=5}`  
`{batch size=1000, number of batches=1}`  
`{batch size=1000, number of batches=3}`  
`{batch size=1000, number of batches=5}`  
**Flat index results:** Time per iteration, memory delta per iteration
**HNSW index results:** Time per iteration,  Recall, memory delta per iteration
2. Variable batch size - Run `Top_K` query where in each iteration the batch size is increased by factor of 2, for each pair of `{batch initial size, number of batches}` from the following:  
`{batch initial size=10, number of batches=2}`  
`{batch initial size=10, number of batches=4}`  
`{batch initial size=100, number of batches=2}`  
`{batch initial size=100, number of batches=4}`  
`{batch initial size=1000, number of batches=2}`  
`{batch initial size=1000, number of batches=4}`  
**Flat index results:** Time per iteration
**HNSW index results:** Time per iteration, Recall, memory delta per iteration

> 3. Batches to Adhoc BF - In each iteration we run `Top_K` with an increasing ` batch size` (initial size=10, increase factor=2) for `number of batches` and then switch to ad-hoc BF. We define `step` as the ratio between the index size to the number of vectors to go over in ad-hoc BF. The tests runs for each pair of `{step, number of batches}` from the following:  

Reviewer 2: I don't think so. this is the benchmarks page

`{step=5, number of batches=0}`  
`{step=5, number of batches=2}`  
`{step=5, number of batches=5}`  
`{step=10, number of batches=0}`  
`{step=10, number of batches=2}`  
`{step=10, number of batches=5}`  
`{step=20, number of batches=0}`  
`{step=20, number of batches=2}`  
`{step=20, number of batches=5}`  
**Flat index results:** Time per iteration
**HNSW index results:** Time per iteration, memory delta per iteration

## BM_VecSimUpdatedIndex

> In this tests set we create 4 indices, 2 of each algorithm (flat and hnsw). The first 2, bf and hnsw indices, contain 500K vectors added to an empty index. The next 2 indices also contain 500K vectors, but this time they were added by overriding the 500K vectors of the aforementioned indices. Currently, we only test this for fp32 single value index.  

Reviewer 2: ```suggestion
In this test set, we create 2 indices for each algorithm. The first index contains 500K vectors added to an empty index. The other index also contains 500K vectors, but this time they were added by overriding the 500K vectors of the aforementioned indices. Currently, we only test this for fp32 single value index.  
```

# The benchmarks directory 
This directory contains the following:
* **Header** files with the classes that are used to define the benchmark test routines. See [Fixtures](#fixtures) section for details.   
It also includes `bm_definitions.h` header that contains benchmarks utils such as:
    - `IndexType` struct
    - common typedefs
    - macro shortcuts to the static data members of `BM_VecSimGeneral` and `BM_VecSim_Index<index_type_t>`
* **subdirectory bm_initialization** - Each file is associated with a test set (basics/batch etc) of a particular **data type**, and fits for both single and multi indices. The purpose  of these files is to avoid code replications. Here we define and register the tests using Google benchmarks macros.  See [How to define and register a new test](#how-to-define-and-register-a-new-test) section.
* **subdirectory run_files** - contains a file for each combination of: set of benchmark tests, index data type, index type(single/multi).  
In this file we:
    - Initialize **BM_VecSimGeneral static data members** such as is_multi, dim, n_vectors, hnsw_index_file, etc, according to the set of parameters of the index loaded from the specified file.

>     - In addition, the macro `BM_FUNC_NAME(bm_func, algo)` is defined to be used in the initialization file. The tests' names are used to generate the graphs on the Grafana dashboard. If two tests share the same name, they will appear on the same graph. Hence, This macro has to **define unique benchmark tests names**. (see [Tests naming](#tests-naming))

Reviewer 2: agree

>     - Include the bm_initialization file that calls google benchmark definitions and registrations macros.

Author: i think its a bit confusing as there is no file called bm_initialization
maybe "Include the initialization file from the `bm_initialization` subdirectory" ?

>     - Include the bm_initialization file that calls google benchmark definitions and registrations macros.

Reviewer: sounds better...

    - Call `BENCHMARK_MAIN()` 

## How to define and register a new test
Google benchmarks library supports templated fixture (see [Google benchmarks user guide](https://github.com/google/benchmark/blob/main/docs/user_guide.md#templated-benchmarks) regarding this topic).  

### Definition:  
`BENCHMARK_TEMPLATE_DEFINE_F(fixture_name, test_name, data type)(benchmark::State &st){fixture_name::test_implementation(st, test_args);}`  
In the definition we:
1. Bind the test to a specific fixture
2. Give it a unique name (see [Tests naming](#tests-naming))
3. Specify the fixture data type (currently available: `fp32_index_t` and `fp64_index_t`) (see [Fixture template type](#fixture-template-type).  
4. Call the function, defined in the fixture class, that implements the required test case.
### Registration
`BENCHMARK_REGISTER_F(test_fixture, test_name)->Args({val1, val2})->ArgNames({"name1", "name2"})->Iterations(val0)->Unit(benchmark::kMillisecond)`  

> Here we register to the benchmarks running loop the test that was bind to `test_fixture` and has `test_name` .  

Reviewer 2: agree

> Here we register to the benchmarks running loop the test that was bind to `test_fixture` and has `test_name` .  

Author: This is about how **we** use the google benchmark macros, and also which of them
To me, it wasn't that trivial ... 
I left the macros and the 4th note
"fixture_name::test_implementation()  should be defined in the fixture class that implements the required test case."

> *NOTE:* The **registration order** is important. Changing it might violate the ability to compare between one benchmark run to another. The reason to that is that we have tests (such as addLabel and DeleteLabel) that modify the hnsw graph.

Reviewer 2: ```suggestion
*NOTE:* The **registration order** is important. Changing it might violate the ability to compare one benchmark run to another. We have tests (such as addLabel and DeleteLabel) that modify the indices and are not idempotent.
```

# Fixtures
## Fixture template type
Since Google benchmarks supports fixture with only one template argument, we use a helper struct to define multiple types.  

> ```asm

Reviewer 2: ```suggestion
``c++
```

Author: cool

// benchmarks/bm_definitions.h
template <VecSimType type, typename DataType, typename DistType = DataType>
struct IndexType {

    static VecSimType get_index_type() { return type; }

    typedef DataType data_t;
    typedef DistType dist_t;
};

// benchmarks/bm_vecsim_index.h
template <typename index_type_t>
class BM_VecSimIndex : public BM_VecSimGeneral {
    
    // Parsing the struct's types
    using data_t = typename index_type_t::data_t;
    using dist_t = typename index_type_t::dist_t;
    ...
    static std::vector<std::vector<data_t>> queries;

    // Using the VecSimType getter 
    static void Initialize() {
        VecSimType type = index_type_t::get_index_type();
        ...
    }
};

```
## Benchmarks fixture classes hierarchy 
You can find the hierarchy graph, defining the main members of each class here:  

> [Benchmarks directory hierarchy graph](https://docs.google.com/presentation/d/1_vqZ50--H9Jv1yf-Pczq30n-kjtO9LW_XYOuH-D0Xjk/edit#slide=id.p).

Reviewer 2: add image

### class BM_VecSimGeneral
Defined in bm_vecsim_general.h  

> This class inherits from includes all the benchmark::Fixture.  

Author: :rofl: i was tired

> This class inherits from includes all the benchmark::Fixture.  

Reviewer 2: use some sort of text proofing tool?

It includes the index parameters as static data members. The static data members are initialized in the .cpp files under benchmark/run_files.  
To refer a member from one of the derived classes, BM_VecSimGeneral:: must be used. Shortcut macros such as `#define N_QUERIES BM_VecSimGeneral::n_queries` are availble in bm_definitions.h.
### Remarks

> * block_size is public because it is used to define the number of iterations on some test cases

Reviewer 2: put this in the code...

> ### class BM_VecSim_Index<index_type_t> : public BM_VecSimGeneral

Reviewer 2: please remove inheritance from the titles, refer to the class and the illustration in the text

Defined in bm_vecsim_index.h  
This class purpose is to initialize the indices and queries data.
It contains:

> * A vector `indices` of `VecSimIndex *` type. One can refer to an index using `VecSimAlgo` enum, where `VecSimAlgo = VecSimAlgo_BF = 0` and `VecSimAlgo = VecSimAlgo_HNSWLIB = 1`.   

Reviewer: Consider saying "a list `indices`..." instead of "vector" for readability

> First, We load a serialized hnsw index from a file, read this index vectors one by one and add the vectors manually to the flat index.  

Reviewer: ```suggestion
In the initialization function, we load serialized HNSW index data from a file and insert the same vectors to the flat index under the same labels.
```

> *NOTE:* The updated tests set is using this vector to save its 4 indices as described in [BM_VecSimUpdatedIndex](#bm_vecsimupdatedindex), where `indices[0]` and `indices[1]` are the BF and HNSW **before** the update, respectively, and `indices[2]` and `indices[3]` are the BF and HNSW **after** the update, respectively.

Reviewer: ```suggestion
*NOTE:* The `BM_VecSimUpdatedIndex` benchmark is using this list to save its four indices as described [above](#bm_vecsimupdatedindex), where `indices[0]` and `indices[1]` are the flat and HNSW **before** the update, respectively, and `indices[2]` and `indices[3]` are the flat and HNSW **after** the update, respectively.
```

> * A vector `queries` contains query vectors of `data_t` type.  

Reviewer 2: ```suggestion
* A list, `queries`, contains query vectors of `data_t` type.  
```

Both Vectors are populated using the files defined in `BM_VecSimGeneral`.
* Reference count - each benchmark definition instantiate a new object of the `test_fixture`. To avoid calling the initialization function multiple times, we use the reference count in the constructor- if it's 0 we initialize the vectors mentioned above, otherwise just increasing the reference count.  

> Using the reference count we also make sure we don't free the indices before we finished all the set's test - the class destructor decreases the reference count, and if it have reached 0 - we free the indices.

Reviewer: It also allows us to save memory (this is the original purpose..)

### Remarks

> * `indices` and `queries` vectors needs to be explicitly initialized for a specific type. For example:

Reviewer 2: ```suggestion
* `indices` and `queries` lists need to be explicitly initialized for a specific type. For example:
```

```asm
template <>
std::vector<std::vector<float>> BM_VecSimIndex<fp32_index_t>::queries{};

template <>
std::vector<VecSimIndex *> BM_VecSimIndex<fp32_index_t>::indices{};
```
### class BM_VecSimCommon<index_type_t> : public BM_VecSimIndex<index_type_t>
Defined in bm_common.h  

> In this class we define the test routines common to the **basics** benchmark and the [**updated index**](#bm_vecsimupdatedindex) benchmarks:

Reviewer: ```suggestion
In this class we define the test routines which are common between the **basics** benchmark and the [**updated index**](#bm_vecsimupdatedindex) benchmarks:
```

* `TopK_BF`
* `TopK_HNSW`
* `Memory_FLAT` 
* `Memory_HNSW`  
### Remarks
These function has `index_offset` parameter (defaults to 0) to be used when we want to apply the function on an updated index.  
In addition, in this file we define macros with the arguments for the `TopK` registration.

### class BM_VecSimBasics<index_type_t> : public BM_VecSimCommon<index_type_t>
Defined in bm_vecsim_basics.h  
In this class we define additional tests for the basic tests set:
* `AddLabel` - Add one label per iteration (meaning 1 vector for single value index, and multiple vector for multi value index)
* `DeleteLabel<algo_t>` - Delete one label per iteration (meaning 1 vector for single value index, and multiple vector for multi value index).  

> `algo_t` is a specific index type (including multi/single) because we call `GetDataByLabel()`, which is only known to HNSW_Single/Multi and BruteForce_Single/Multi. This why this test case is defined separately in each .cpp file using `DEFINE_DELETE_LABEL` macro (defined in bm_vecsim_basics.h). The registration is done like all other test cases in the initialization file.  

Reviewer: ```suggestion
`algo_t` is the specific index type that is required to obtain vector data by its label (used to restore the deleted vectors). This requirement for `algo_t` parameter is the reason that this test case is defined separately in each .cpp file using `DEFINE_DELETE_LABEL` macro (defined in bm_vecsim_basics.h). The registration is done like all other test cases in the initialization file.  
```

* `Range_BF` 
* `Range_HNSW`  
In addition, in this file we define macros with the arguments for the `Range`, `AddLabel` and `DeleteLabel` registration.
### Remarks
The [basic benchmark set](#bm_vecsimbasics) includes functions defined both in `BM_VecSimCommon` and `BM_VecSimBasics`. Pay attention to refer the correct class when defining the test in the [initialization file](#how-to-define-and-register-a-new-test).
### class BM_VecSimUpdatedIndex<index_type_t> : public BM_VecSimCommon<index_type_t>
Defined in bm_updated_index.h  
In this class constructor we add the additional indices, the "updated" indices. The "before" indices are initialized in `BM_VecSimIndex`.  
This class defines no new test routines and uses only tests defined in `BM_VecSimCommon`.  
### Remarks
The constructor is called after we already defined the tests resides in `BM_VecSimCommon`, so `ref_count` is not zero and we cant rely on it to decide weather we should initialize the indices or not. This is why we have to hold another the `is_initialized` flag.  

> Also, we keep the value of `ref_count` at the moment of initialization in `first_updatedBM_ref_count` to free the indices when `ref_count` decreased to this value.

Reviewer: Not clear to me...

Author: moved to code (bm_updated_index.h) lmk if it's better explained now

> Also, we keep the value of `ref_count` at the moment of initialization in `first_updatedBM_ref_count` to free the indices when `ref_count` decreased to this value.

Reviewer 2: this is TMI

Author: K moving it to the code

### class BM_BatchIterator<index_type_t> : public BM_VecSimIndex<index_type_t>
Defined in bm_batch_iterator.h  
Here we implement the tests routines of [batch iterator set](#bm_batchiterator) to be called by the test definition in the [initialization file](#how-to-define-and-register-a-new-test).  
In this file we also define the arguments for each test case in the batch iterator tests set.

### Tests counters
* Tests counters finale values are printed in the benchmarks output next to the test results. 
* They can also be used as input data in the Grafana dashboard.
### Tests naming:
A test name will look something like this:  
`test_fixture<fixture_data_type>/test_name/arg1:val/iterations:val`  
for example:  
`BM_VecSimBasics<fp32_index_t>/Range_HNSW_Single/radiusX100:20/epsilonX1000:1/iterations:100`  
So test's name components are:
1. The test's fixture and the fixture type (if it's templated)
2. The name as given at definition and registration
3. arguments names and values  
*NOTE:* the fixture type (fp32_index_t, for example) is part of the name, so tests that use different fixture **or** the same fixture of a different type, **can** share the same name.  

> # Cmake remarks

Reviewer 2: you can remove this

Author: this = the note + cmake remarks section?

Each test binary includes only one file .cpp from benchmark/run_files because
1. We don't want to override BM_VecSimGeneral static data members

> 2. Memory (RAM) limitations.

Reviewer: This is redundant in my opinion... it can be easily deduced from running the benchmarks + explanations above

Author: i removed Tests counters, but I think it won't do any harm to leave the naming section (especially pointing out that the test fixture type also gives the name its uniqueness)

Reviewer: yes

## benchmarks.sh
In this file we define the names of the benchmarks binary files, **without** the `bm_` prefix. The names are used in:
1. cmake of the benchmarks - to define the executable file name (bm_${benchmark}), refer the .cpp files, 
2. Makefile - to execute the benchmarks (with the names defined the cmake) and define the results files names (${benchmark}_results).
3. benchmark.yml - read the results from the files according to the name defined in the Makefile.

# ANN-Benchmark

[ANN-Benchmarks](http://ann-benchmarks.com/) is a benchmarking environment for approximate nearest neighbor algorithms search (for additional information, refer to the project's [github repository](https://github.com/erikbern/ann-benchmarks)).  Each algorithm is benchmarked on pre-generated commonly use datasets (in HDF5 formats).

> The `bm_dataset.py` script uses some of ANN-Benchmark datasets to measure this library performance in the same manner. The following datasets are downloaded and benchmarked:

Reviewer: ```suggestion
The `bm_dataset.py` script uses some of ANN-Benchmark datasets to measure this library performance in the same manner. The following datasets are downloaded and benchmarked (all use FP32 single value per label data):
```

## BM_VecSimBasics
For each combination of data type (fp32/fp64) and index type (single/multi) the following test cases are included:
1. Mesure index total `memory` (runtime and iterations number are irrelevant for this use case, just the memory metric) 
2. `AddLabel` - runs for `DEFAULT_BLOCK_SIZE (= 1024)` iterations, in each we add one new label to the index from the `queries` list. Note that for a single value index each label contains one vector, meaning that the number of new labels equals the number of new vectors.  
**results:** average time per label, average memory addition per vector, vectors per label.  
*At the end of the benchmark, we delete the added labels*
3. `DeleteLabel` - runs for `DEFAULT_BLOCK_SIZE (= 1024)` iterations, in each we delete one label from the index. Note that for a single value index each label contains one vector, meaning that the number of deleted labels equals the number of deleted vectors.  
**results:** average time per label, average memory addition per vector (a negative value means that the memory has decreased).  

> *At the end of the benchmark, we re-add the deleted vectors under the same labels*

Reviewer 2: ```suggestion
*At the end of the benchmark, we restore the deleted vectors under the same labels*
```
